src/ragger/navigator/navigator.py

```diff
@@ -283,3 +283,59 @@ def navigate_until_snap(self,
         else:
             raise ValueError(f"Could not find first snapshot {first_golden_snap}")
         return img_idx
+
+    def navigate_until_text(self,
+                            navigate_instruction: NavIns,
+                            validation_instruction: NavIns,
+                            text: str,
+                            timeout: int = 30):
+        """
+        Navigate until some text is found on the screen content displayed.
+
+        This method may be left void on backends connecting to physical devices,
+        where a physical interaction must be performed instead.
+        This will prevent the instrumentation to fail (the void method won't
+        raise `NotImplementedError`), but the instrumentation flow will probably
+        get stuck (on further call to `receive` for instance) until the expected
+        action is performed on the device.
+
+        :param navigate_instruction: Navigation instruction to be performed until the text is found.
+        :type navigate_instruction: NavIns
+        :param validation_instruction: Navigation instruction to be performed once the text is found.
+        :type validation_instruction: NavIns
+        :param first_instruction_wait: Sleeping time before the first snapshot
+        :type first_instruction_wait: float
+        :param timeout: Timeout of the navigation loop if last snapshot is not found.
+        :type timeout: int
+
+        :return: img_idx
+        :rtype: int
+        """
+
+        if not isinstance(self._backend, SpeculosBackend):
+            # When not using Speculos backend, taking snapshots is not possible
+            # therefore comparison is not possible too.
+            # TODO request user to interact with the device.
+            return 0
+
+        start = time()
+
+        ctx = self._backend.wait_for_screen_change(2.0)
+
+        # Navigate until the text specified in argument is found.
+        while True:
+            now = time()
+
+            ctx = self._backend.wait_for_screen_change(2.0, ctx)
```

Comment:
However this one will probably raised in case `navigate_until_text` is called after sending an APDU which flow is quickly displayed.
Something like :
- Send APDU
- Flow displayed
- `navigate_until_text`:
    - first wait_for_screen_change(2.0) => get first screen of flow content
    - wait_for_screen_change(2.0) in the while loop => timeout as no navigation instruction was sent.

Reply:
I moved the navigation instruction above the wait for screen change in the loop. It should solve the problem you describe (and other use cases with no initial automatic screen change).